gitlab/client.py

```diff
@@ -749,7 +749,7 @@ def http_request(
             self._check_redirects(result.response)
 
             if 200 <= result.status_code < 300:
-                return result.response
+                return result
```

Comment:
This is a breaking change, so this would need to be at least communicated to users via a breaking change trailer (and triger 4.0.0). But also I would like to see how this affects our streamed responses and how we deal with `download()` when `iterator=True` is passed to it, where people iterate over chunks on their own.

I'd like to do some local testing for this to see what behavior we get.

Reply:
Make sense. How can I help?

Reply:
@lmilbaum I like your idea with protocols (or maybe abstract classes) in the other PR, but I would maybe try to apply it here for backends instead of the client, if that makes sense. I actually had a similar WIP locally, where I started with some ABC work on the backends rather than the client itself. Have you tried that as well?

Reply:
@nejch This is a great idea. Updated the other PR. Take a look.